src/main/java/seedu/address/model/VersionedAddressBook.java

```diff
@@ -15,31 +15,63 @@ public class VersionedAddressBook extends AddressBook {
     // Keeps track of the current version of AddressBook
     private int currentStatePointer;
 
+    // Maximum number of snapshots we take
+    private int sizeLimit;
+
     // Stores the different versions of AddressBook
     private ArrayList<ReadOnlyAddressBook> addressBookStateList = new ArrayList<>();
 
     /**
      * Creates an instance of VersionedAddressBook.
+     *
      * @param addressBook AddressBook to be initialised as the first state
      */
     public VersionedAddressBook(ReadOnlyAddressBook addressBook) {
         this.currentStatePointer = 0;
+        this.sizeLimit = 20; // Default value
         addressBookStateList.add(this.currentStatePointer, addressBook);
     }
 
+    /**
+     * Creates an instance of VersionedAddressBook.
+     *
+     * @param addressBook AddressBook to be initialised as the first state
+     * @param stepLimit   maximum number of previous states tracked
+     */
+    public VersionedAddressBook(ReadOnlyAddressBook addressBook, int stepLimit) {
+        this.currentStatePointer = 0;
+        this.sizeLimit = stepLimit; // Default value
+        addressBookStateList.add(this.currentStatePointer, addressBook);
+    }
+
+    private void trimStateList() {
+        if (this.addressBookStateList.size() <= this.sizeLimit) {
+            return;
+        }
+        while (this.addressBookStateList.size() > this.sizeLimit) {
+            this.addressBookStateList.remove(0);
+            this.currentStatePointer--;
+        }
+    }
+
```

Comment:
I think the new addition of setting the number of undo to 20 by default and also giving the user the ability to set the limit is good. I like that the list will trim and remove the oldest state when it exceeds the default value.